temporian/implementation/numpy_cc/operators/window.cc

```diff
@@ -10,436 +10,568 @@
+
+    while (end_idx < n_event && v_timestamps[end_idx] <= right_limit) {
+      accumulator.Add(v_values[end_idx]);
+      end_idx++;
     }
 
-    return output;
+    while (begin_idx < n_event &&
+           // Compare both sides around ~0 to get maximum float resolution
+           v_sampling[sampling_idx] - v_timestamps[begin_idx] >=
+               window_length) {
+      accumulator.Remove(v_values[begin_idx]);
+      begin_idx++;
+    }
+
+    v_output[sampling_idx] = accumulator.Result();
   }
 
-  // Note: We only use inheritence to compile check the code.
-  template <typename INPUT, typename OUTPUT>
-  struct Accumulator
-  {
-    virtual ~Accumulator() = default;
-    virtual void Add(INPUT value) = 0;
-    virtual void Remove(INPUT value) = 0;
-    virtual OUTPUT Result() = 0;
-  };
-
-  template <typename INPUT, typename OUTPUT>
-  struct SimpleMovingAverageAccumulator : Accumulator<INPUT, OUTPUT>
-  {
-    void Add(INPUT value) override
-    {
-      if constexpr (std::numeric_limits<INPUT>::has_quiet_NaN)
-      {
-        if (std::isnan(value))
-        {
-          return;
+  return output;
+}
+
+// No external sampling, variable window length
+template <typename INPUT, typename OUTPUT, typename TAccumulator>
+py::array_t<OUTPUT> accumulate(const ArrayD &evset_timestamps,
+                               const py::array_t<INPUT> &evset_values,
+                               const ArrayD &window_length) {
+  // Input size
+  const size_t n_event = evset_timestamps.shape(0);
+
+  // Allocate output array
+  auto output = py::array_t<OUTPUT>(n_event);
+
+  auto v_output = output.template mutable_unchecked<1>();
+  auto v_timestamps = evset_timestamps.unchecked<1>();
+  auto v_values = evset_values.template unchecked<1>();
+  auto v_window_length = window_length.unchecked<1>();
+
+  TAccumulator accumulator;
+
+  // Index of the first value in the window.
+  size_t begin_idx = 0;
+  // Index of the first value outside the window.
+  size_t end_idx = 0;
+
+  double curr_ts;
+  double curr_window_length;
+  size_t first_diff_ts_idx;
+
+  while (end_idx < n_event) {
+    // Note: We accumulate values in (t-window_length, t] with t=
+    // v_timestamps[end_idx], and there may be several contiguous equal
+    // values in v_timestamps.
+
+    curr_ts = v_timestamps[end_idx];
+    // TODO: raise if negative
+    curr_window_length = v_window_length[end_idx];
+
+    // Add all values with same timestamp as the current one.
+    accumulator.Add(v_values[end_idx]);
+    first_diff_ts_idx = end_idx + 1;
+    while (first_diff_ts_idx < n_event &&
+           v_timestamps[first_diff_ts_idx] == curr_ts) {
+      accumulator.Add(v_values[first_diff_ts_idx]);
+      first_diff_ts_idx++;
+    }
+
+    if (end_idx > 0) {
+      // Move begin_idx forward or backwards depending on begin_diff.
+      if (curr_ts - v_timestamps[end_idx - 1] -
+              (curr_window_length - v_window_length[end_idx - 1]) >
+          0) {

```

Comment:
Nice, I think this ensures that it works with maximum float precision.
It would be great if we can add a test using `tp.duration.shortest` (see `test_cc_max_resolution_wo_sampling` in moving_count_test) to ensure that it works here as well.

Reply:
added